### ISSUE TITLE
routes to external resources

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#2016]
extend the image widget to interpret its source as relative to a **route** (/ path / base url) read from a global config tiddler that specifies the location, e.g...

```
title: $:/config/routes/default

http:/mysite.com/img/
```

with...

```
<$image source="foo.jpg">
```

...automatically switching to using the **default** route when defined, whereas...

```
<$image source="http://foo.com/bar/baz.jpg">
```

...automatically detects via `://` that we're dealing with a fully qualified url, and thus ignores any routing.

Therefore, if defined, the **default** route would always be implied with any $image widget pulling external contents. Alternatively, the user could prevent a default route via `route=""` and thus use a truly relative path.

There can be a number of routes like so...

```
title: $:/config/routes/bucket

http://photobucket.com/user/myusername/mywiki/
```

with...

```
<$image route="bucket" source="works.jpg">
```

and then...

```
[img route=bucket[works.jpg]]
```

The core should be fine with all this as it doesn't rely on external resources for any of its ui / components.

**Comment by Jermolene** [comment]
Hi @tobibeer are you familiar with the operation of the variable `tv-get-export-image-link`?

https://github.com/Jermolene/TiddlyWiki5/search?utf8=%E2%9C%93&q=tv-get-export-image-link
https://github.com/Jermolene/TiddlyWiki5/pull/1509

Doesn't it offer rather similar functionality?

**Comment by tobibeer** [comment]
Not familiar with [tv-get-export-image-link](http://tiddlywiki.com/prerelease/#tv-get-export-image-link%20Variable). So, this provides a way address external images like so...

```
title: $:/config/routes/default
tags: $:/tags/Macro

\define tv-get-export-image-link(src)
http://my-server.com/img/$src$
\end
```

And then...

```
[img[foo.jpg]]
```

...in any tiddler, without having to specifically tell the widget not to use the route.

~

I see a few significant problems with this "variable approach", I guess in general:

1) Fully qualified URLs will break:

```
[img[http://my-other-server.com/img/bar.jpg]] 
```

2) I need to declare another complexish macro for each tiddler that wants to point somewhere else, e.g.

```
title: some tiddler not using the default external path

\define tv-get-export-image-link(src)
http://my-other-server.com/img/$src$
\end

[img[bar.jpg]]
[img[baz.jpg]]
```

3) I can't (easily) do 2) whenever a single tiddler would want to make use of two different routes.

~

I think I'm not too fond of using these rather cryptic variable names as compared to short memorable syntax.

```
[img route=foo[bar.jpg]
<$image route="my-named-route" source="bar.jpg"/>
```

Makes it all so much more manageable, also to define new routes in a simple ui, immediately available without any further macro fiddling.

**Comment by tobibeer** [comment]
As commented [here](https://groups.google.com/d/msg/tiddlywiki/lLHcVn0MRNI/uztf_qywDgAJ)...

> It appears that it is not a widget that needs to specify what **route** an image tiddler specifying a **_canonical_uri** is to use — or, perhaps only as an override — but rather that it is the image tiddler itself that needs to qualify the **route** / base-url / path to which it is be resolved, unless qualified with a full url (at which point an externalizing process should never touch them).

Then, to point to a different location all that needs chaning is the tiddler that configures the path for any given route.

**Comment by pmario** [comment]
TLDR;
I think, this is plugin territory. The core needs to provide basic functionality and must not block or hinder plugin authors, if they need much more freedom.

---------

I think your approach is one possible solution, that only solves the problem partially. 

IMO `<$image route="bucket" source="works.jpg">` is a good idea, but will be not enough freedom !

As a user I want to use `[img[bar.jpg]` and not `[img route=foo[bar.jpg]` but I still want all the possibilities like with fancybox described below. 

eg:

"file://" and "http://" image paths and activating a different behaviour needs so much more configuration, than you pointed out here. see my [fancybox](http://fancybox.tiddlyspace.com/) project for TWC. 

 - sensible fallbacks are needed, if eg: small images are embedded in the TW and the full size image is linked with _cannonical_uri. ...
 - Config parameters I think are needed: http://fancybox.tiddlyspace.com/#FancyBo - PluginInfo
 - Per tiddler configuration options: http://fancybox.tiddlyspace.com/#13.jpg

**Comment by tobibeer** [comment]
@pmario, I was actually suggesting that when a default route was defined that widgets as currently used would not need any route-specifying, but rather that it would be applied by default, so all wikitext image references with relative paths for resources stay short and sweet.

> "file://" and "http://" image paths and activating a different behaviour needs so much more configuration

how / why? one config tiddler (per route) that points to where to find things

> sensible fallbacks are needed, if eg: small images are embedded in the TW and the full size image is linked with _cannonical_uri

that's what routes are for ...besides, they only apply to external images, not embedded thumbnails

> config parameters I think are needed

I think you're confusing using and configuring image tiddlers (embedded or external) with configuring any plugins that subsequently use them, those appear to be two different things

> per tiddler configuration options

all this can be done with fields, and only applies to whichever fancy-plugin someone wishes to use to display images in more sophisticated ways

**Comment by pmario** [comment]
>To cut things short: This is not about some FancyBox-type plugin (!)

That's exactly the point. You think my examples are wrong for the core. And yes they are, because they are for a plugin!

I think your proposals behave like a plugin.

eg:

>```
[img route=bucket[works.jpg]]
```

If this is part of the core, we will have to stick with this syntax for the rest of the TW lifetime. ... As I wrote, I like the idea and the abstraction, but I think it only covers a part of the problem. ... 

I don't want to write `[img route=bucket[works.jpg]]`. I want to write `[img[works.jpg]]` and still have the functionality you describe. I know it is possible, because I've done it with TWc. 

So a new proposal like yours should be implemented as a plugin first. So we can see, if it works and what can be improved. ... After several iterations we may have something that is ready for the core.

**Comment by tobibeer** [comment]
At some point TiddlyWiki started off with a standalone wiki. Turns out, things have changed... and so it is important that we cater for a few more ideas that are non-standalone, in the core.

To me — "routing" or whatever you want to call it — is a core feature in most any environment that allows access to stuff located somewhere else.

> I think your proposals behave like a plugin.

So, I don't agree. Here is a situation with a wiki managed via node, publishing as a standalone or static version someplace else, with links to resources breaking. I think that finding a simple solution to this deserves a "core" badge, whatever the solution may be.

> [img route=bucket[works.jpg]]
> If this is part of the core, we will have to stick with this syntax for the rest of the TW lifetime.
> I want to write [img[works.jpg]] and still have the functionality you describe.

I can only repeat: the basic proposal works precisely like that! The `route="foo"` syntax is only icing on the cake. It is not essentially needed for the proposed functionality. All it would allow is to define not just one external path but multiple ones and that appears something I believe people will use rather often.

In other words, if we have a mechanism to define one route and don't use the opportunity to allow for multiple routes — or whatever the name — then we're doing it wrong.

> So a new proposal like yours should be implemented as a plugin first. So we can see, if it works and what can be improved. ... After several iterations we may have something that is ready for the core.

Just as a reminder. Doing conceptual work and discussion implementation like this by no means reads as "Jeremy, please do this, pronto."

**Comment by Jermolene** [comment]
I'm also not persuaded that this belongs in the core, in any case the discussion is somewhat speculative until we have an implementation.

I'm closing the ticket on the basis that it is not actionable at present.
